lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -859,6 +859,31 @@ public function iterate($parameters = null, $hydrationMode = null)
         return $this->em->newHydrator($this->hydrationMode)->iterate($stmt, $rsm, $this->hints);
     }
 
+    /**
+     * Executes the query and returns an iterable that can be used to incrementally
+     * iterate over the result.
+     *
+     * @param ArrayCollection|mixed[]|null $parameters    The query parameters.
+     * @param string|int|null              $hydrationMode The hydration mode to use.
+     *
+     * @return iterable<mixed>
+     */
+    public function getIterable($parameters = null, $hydrationMode = null) : iterable
```

Comment:
I think a type declaration will not hurt:

```suggestion
    public function getIterable(?iterable $parameters = null, $hydrationMode = null) : iterable
```

Reply:
Unfortunately it would not comply with `setParameters()` that accepts `ArrayCollection|mixed[]` so `iterable` is no covariant with it. Or am I missing something?

Reply:
I think `ArrayCollection|mixed[]` is iterable, isn't it? So it's compatible, I think…

Reply:
This could be changed to work with `iterable` easily:

```diff
-       if ($parameters !== null && count($parameters) !== 0) {
+       if ($parameters !== null) {
            $this->setParameters($parameters);
        }
```

But I think that if I pass eg. something implementing `Iterator` it would not comply with `ArrayCollection<Parameter>` as it does not have its methods (`clear()` for example).

```php
    /**
     * The parameter map of this query.
     *
     * @var ArrayCollection|Parameter[]
     */
    protected $parameters;
```

Reply:
Nothing tells you that an iterable that is passed will not be countable, so there is nothing to change here IMO. Using `iterable` is not a sufficient contract for using `count()`, but it's better than no type declaration at all I think.

Reply:
You can't `count()` `iterable`. For that I'd also have to check for `is_countable()` but that's available since 7.3 so it's not applicable anyway. 

But still, the main issue is incompatibility between `ArrayCollection` and `iterable`. You cannot assign `iterable` to `ArrayCollection`.

https://phpstan.org/r/90dedd75-2907-46fc-897e-1d368d8dc75c

```php
<?php

declare(strict_types=1);

/**
 * @template K
 * @template V
 */
final class X
{
    /** @var ArrayIterator<K, V> */
    private $iterator;

    public function __construct()
    {
        $this->iterator = new ArrayIterator([]);
    }

    /** @param iterable<K, V> $x */
    public function setIterable(iterable $x) : void
    {
        $this->iterator = $x;
    }
}
```

Reply:
I was not suggesting you change the phpdoc, I was suggesting you do just this, which phpstan is fully OK with: https://phpstan.org/r/c20675d0-8b8c-4ba6-8ca3-5812bb371265

Reply:
Ah ok, now it clicked. Thanks